File: results/0.0/PNG-Image/png-image-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define CHUNK_SIZE 1024

typedef struct {
    uint32_t width;
    uint32_t height;
    uint8_t bit_depth;
    uint8_t color_type;
    uint8_t compression_method;
    uint8_t filter_method;
    uint8_t interlace_method;
} ihdr_chunk_t;

typedef struct {
    uint8_t red;
    uint8_t green;
    uint8_t blue;
} plte_entry_t;

typedef struct {
    plte_entry_t entries[256];
    uint8_t num_entries;
} plte_chunk_t;

typedef struct {
    uint8_t* data;
    uint32_t size;
} idat_chunk_t;

typedef struct {
    uint32_t crc;
} iend_chunk_t;

typedef struct {
    char type[5];
    uint8_t* data;
    uint32_t size;
    uint32_t crc;
} chunk_t;

typedef struct {
    uint8_t magic_number[8];
    ihdr_chunk_t ihdr;
    plte_chunk_t plte;
    idat_chunk_t idat;
    iend_chunk_t iend;
    chunk_t* chunks;
    uint32_t num_chunks;
} png_image_t;

void* png_magic_number_parser(void* input) {
    static uint8_t magic_number[8] = {0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A};
    if (memcmp(input, magic_number, 8) == 0) {
        return input + 8;
    } else {
        return NULL;
    }
}

void* ihdr_chunk_parser(void* input) {
    ihdr_chunk_t* ihdr = (ihdr_chunk_t*)input;
    input += sizeof(ihdr_chunk_t);
    return input;
}

void* plte_chunk_parser(void* input) {
    plte_chunk_t* plte = (plte_chunk_t*)input;
    input += sizeof(plte_chunk_t);
    return input;
}

void* idat_chunk_parser(void* input) {
    idat_chunk_t* idat = (idat_chunk_t*)input;
    input += sizeof(idat_chunk_t);
    return input;
}

void* iend_chunk_parser(void* input) {
    iend_chunk_t* iend = (iend_chunk_t*)input;
    input += sizeof(iend_chunk_t);
    return input;
}

void* chunk_parser(void* input) {
    chunk_t* chunk = (chunk_t*)input;
    input += sizeof(chunk_t);
    return input;
}

void* png_image_parser(void* input) {
    png_image_t* image = (png_image_t*)input;
    input = png_magic_number_parser(input);
    if (input == NULL) {
        return NULL;
    }
    input = ihdr_chunk_parser(input);
    if (input == NULL) {
        return NULL;
    }
    input = plte_chunk_parser(input);
    if (input == NULL) {
        return NULL;
    }
    input = idat_chunk_parser(input);
    if (input == NULL) {
        return NULL;
    }
    input = iend_chunk_parser(input);
    if (input == NULL) {
        return NULL;
    }
    return input;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        printf("Error opening file: %s\n", argv[1]);
        return 1;
    }

    png_image_t image;
    void* input = &image;
    void* result = png_image_parser(input);

    if (result != NULL) {
        printf("PNG image parsed successfully:\n");
        printf("Width: %u\n", image.ihdr.width);
        printf("Height: %u\n", image.ihdr.height);
        printf("Bit depth: %u\n", image.ihdr.bit_depth);
        printf("Color type: %u\n", image.ihdr.color_type);
        printf("Compression method: %u\n", image.ihdr.compression_method);
        printf("Filter method: %u\n", image.ihdr.filter_method);
        printf("Interlace method: %u\n", image.ihdr.interlace_method);
        if (image.plte.num_entries > 0) {
            printf("PLTE chunk:\n");
            for (uint8_t i = 0; i < image.plte.num_entries; i++) {
                printf("Entry %u: (%u, %u, %u)\n", i, image.plte.entries[i].red, image.plte.entries[i].green, image.plte.entries[i].blue);
            }
        }
        if (image.idat.size > 0) {
            printf("IDAT chunk:\n");
            printf("Size: %u\n", image.idat.size);
        }
        printf("IEND chunk:\n");
        printf("CRC: %u\n", image.iend.crc);
    } else {
        printf("Error parsing PNG image\n");
    }

    fclose(file);
    return 0;
}

File: results/0.0/PNG-Image/png-image-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{tag},
    multi::{length_data},
    number::complete::{be_u32, be_u8},
    IResult,
};
use std::{
    env,
    fs::File,
    io::{BufReader, Read},
};
use std::path::Path;

#[derive(Debug, PartialEq)]
enum ColorType {
    Grayscale,
    RGB,
    Plte,
    GreyscaleAlpha,
    RGBA,
}

#[derive(Debug, PartialEq)]
enum CompressionMethod {
    DeflateInflate,
}

#[derive(Debug, PartialEq)]
enum FilterMethod {
    Adaptive,
}

#[derive(Debug, PartialEq)]
enum InterlaceMethod {
    NoInterlace,
    Adam7Interlace,
}

#[derive(Debug, PartialEq)]
struct IHDR {
    width: u32,
    height: u32,
    bit_depth: u8,
    color_type: ColorType,
    compression_method: CompressionMethod,
    filter_method: FilterMethod,
    interlace_method: InterlaceMethod,
}

fn parse_magic_number(input: &[u8]) -> IResult<&[u8], ()> {
    static MAGIC_NUMBER: [u8; 8] = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];
    tag(&MAGIC_NUMBER)(input).map(|(input, _)| (input, ()))
}

fn parse_ihdr(input: &[u8]) -> IResult<&[u8], IHDR> {
    let (input, _) = tag([0x49, 0x48, 0x44, 0x52])(input)?;
    let (input, width) = be_u32(input)?;
    let (input, height) = be_u32(input)?;
    let (input, bit_depth) = be_u8(input)?;
    let (input, color_type) = be_u8(input)?;
    let (input, compression_method) = be_u8(input)?;
    let (input, filter_method) = be_u8(input)?;
    let (input, interlace_method) = be_u8(input)?;
    Ok((
        input,
        IHDR {
            width,
            height,
            bit_depth,
            color_type: match color_type {
                0 => ColorType::Grayscale,
                2 => ColorType::RGB,
                3 => ColorType::Plte,
                4 => ColorType::GreyscaleAlpha,
                6 => ColorType::RGBA,
                _ => panic!("Invalid color type"),
            },
            compression_method: match compression_method {
                0 => CompressionMethod::DeflateInflate,
                _ => panic!("Invalid compression method"),
            },
            filter_method: match filter_method {
                0 => FilterMethod::Adaptive,
                _ => panic!("Invalid filter method"),
            },
            interlace_method: match interlace_method {
                0 => InterlaceMethod::NoInterlace,
                1 => InterlaceMethod::Adam7Interlace,
                _ => panic!("Invalid interlace method"),
            },
        },
    ))
}

fn parse_plte(input: &[u8]) -> IResult<&[u8], Vec<(u8, u8, u8)>> {
    let (input, _) = tag([0x50, 0x4c, 0x54, 0x45])(input)?;
    let (input, palette_data) = length_data(be_u32)(input)?;
    let palette_entries: Vec<(u8, u8, u8)> = palette_data
        .chunks(3)
        .map(|chunk| (chunk[0], chunk[1], chunk[2]))
        .collect();
    Ok((input, palette_entries))
}

fn parse_idat(input: &[u8]) -> IResult<&[u8], Vec<u8>> {
    let (input, _) = tag([0x49, 0x44, 0x41, 0x54])(input)?;
    let (input, idat_data) = length_data(be_u32)(input)?;
    Ok((input, idat_data.to_vec()))
}

fn parse_iend(input: &[u8]) -> IResult<&[u8], ()> {
    tag([0x49, 0x45, 0x4e, 0x44])(input).map(|(input, _)| (input, ()))
}

fn parse_png(input: &[u8]) -> IResult<&[u8], (IHDR, Vec<(u8, u8, u8)>, Vec<u8>)> {
    let (input, _) = parse_magic_number(input)?;
    let (input, ihdr) = parse_ihdr(input)?;
    let (input, plte) = parse_plte(input)?;
    let (input, idat) = parse_idat(input)?;
    let (input, _) = parse_iend(input)?;
    Ok((input, (ihdr, plte, idat)))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <input_file>", args[0]);
        return;
    }
    let input_file = &args[1];
    let path = Path::new(input_file);
    let file = File::open(path).unwrap();
    let mut reader = BufReader::new(file);
    let mut input = Vec::new();
    reader.read_to_end(&mut input).unwrap();
    match parse_png(&input) {
        Ok((remaining, (ihdr, plte, idat))) => {
            println!("IHDR: {:?}", ihdr);
            println!("PLTE: {:?}", plte);
            println!("IDAT: {:?}", idat);
            println!("Remaining: {:?}", remaining);
        }
        Err(err) => {
            println!("Error: {:?}", err);
        }
    }
}